src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,564 @@
+#include "subquery_to_join_rule.hpp"
+
+#include <map>
+#include <memory>
+#include <set>
+
+#include "cost_model/cost_model_logical.hpp"
+#include "expression/abstract_expression.hpp"
+#include "expression/abstract_predicate_expression.hpp"
+#include "expression/binary_predicate_expression.hpp"
+#include "expression/expression_functional.hpp"
+#include "expression/expression_utils.hpp"
+#include "expression/in_expression.hpp"
+#include "expression/lqp_column_expression.hpp"
+#include "expression/lqp_subquery_expression.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/aggregate_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "statistics/table_statistics.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+/**
+ * Info about a predicate which needs to be pulled up into a join predicate.
+ */
+struct PredicateInfo {
+  /**
+   * Comparison operand from the left sub-tree.
+   */
+  std::shared_ptr<AbstractExpression> left_operand;
+
+  /**
+   * Comparison operand from the right sub-tree (the previous subquery).
+   */
+  std::shared_ptr<AbstractExpression> right_operand;
+
+  PredicateCondition condition;
+};
+
+/**
+ * Collected information about the predicates which need to be pulled up and other nodes which need to be adjusted.
+ */
+struct PredicatePullUpInfo {
+  /**
+   * Predicates to pull up.
+   *
+   * Ordered by depth in the sub-tree, from top to bottom.
+   */
+  std::vector<PredicateInfo> predicates;
+
+  /**
+   * Nodes from which the pull up predicates originate.
+   */
+  std::set<std::shared_ptr<PredicateNode>> predicate_nodes;
+
+  /**
+   * Projection nodes in the sub-tree which need to be removed to make the predicate pull-up safe.
+   */
+  std::vector<std::shared_ptr<ProjectionNode>> projection_nodes;
+
+  /**
+   * Alias nodes in the sub-tree which need to be extended to make the predicate pull-up safe.
+   *
+   * The nodes are ordered top to bottom.
+   */
+  std::vector<std::shared_ptr<AliasNode>> alias_nodes;
+
+  /**
+   * Aggregates which need to be patched allow the predicates to be pulled up.
+   *
+   * Ordered by depth in the sub-tree, from top to bottom. The number notes the number of predicates above this
+   * aggregate in the tree, which won't need be considered when patching the aggregate.
+   */
+  std::vector<std::pair<std::shared_ptr<AggregateNode>, size_t>> aggregate_nodes;
+};
+
+/**
+ * Checks whether a predicate node should be turned into a join predicate.
+ */
+std::optional<PredicateInfo> should_become_join_predicate(
+    const std::shared_ptr<PredicateNode>& predicate_node,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping, bool is_below_aggregate) {
```

Comment:
Not necessarily related to this line: LQPs can contain multiple aggregates - is that something that is correctly covered?
Example: `SELECT AVG(x) FROM (SELECT SUM(f) AS x FROM t GROUP BY i) t2`

Reply:
Multiple aggregates are supported, yes. For example, the following query gets reformulated correctly:
```sql
SELECT *
FROM customer
WHERE c_acctbal > (
  SELECT SUM(a.x)
  FROM (
    SELECT AVG(o_totalprice) AS x
    FROM orders
    WHERE o_custkey = c_custkey
    GROUP BY o_orderpriority
  ) AS a
)
```
We will also have tests for cases like this very soon.